File: akka-sample-cqrs-scala/src/main/scala/sample/cqrs/ShoppingCartRoutes.scala
package sample.cqrs

import scala.concurrent.Future
import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.http.scaladsl.model.StatusCodes
import akka.http.scaladsl.server.Route
import akka.pattern.StatusReply
import akka.util.Timeout

object ShoppingCartRoutes {
  final case class AddItem(cartId: String, itemId: String, quantity: Int)
  final case class UpdateItem(cartId: String, itemId: String, quantity: Int)
}

class ShoppingCartRoutes()(implicit system: ActorSystem[_]) {

  implicit private val timeout: Timeout =
    Timeout.create(system.settings.config.getDuration("shopping.askTimeout"))
  private val sharding = ClusterSharding(system)

  import ShoppingCartRoutes._
  import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
  import akka.http.scaladsl.server.Directives._
  import JsonFormats._

  val shopping: Route =
    pathPrefix("shopping") {
      pathPrefix("carts") {
        concat(
          post {
            entity(as[AddItem]) {
              data =>
                val entityRef =
                  sharding.entityRefFor(ShoppingCart.EntityKey, data.cartId)
                val reply: Future[StatusReply[ShoppingCart.Summary]] =
                  entityRef.ask(ShoppingCart.AddItem(data.itemId, data.quantity, _))
                onSuccess(reply) {
                  case StatusReply.Success(summary: ShoppingCart.Summary) =>
                    complete(StatusCodes.OK -> summary)
                  case StatusReply.Error(reason) =>
                    complete(StatusCodes.BadRequest -> reason)
                }
            }
          },
          put {
            entity(as[UpdateItem]) {
              data =>
                val entityRef =
                  sharding.entityRefFor(ShoppingCart.EntityKey, data.cartId)

                def command(replyTo: ActorRef[StatusReply[ShoppingCart.Summary]]) =
                  if (data.quantity == 0)
                    ShoppingCart.RemoveItem(data.itemId, replyTo)
                  else
                    ShoppingCart.AdjustItemQuantity(data.itemId, data.quantity, replyTo)

                val reply: Future[StatusReply[ShoppingCart.Summary]] =
                  entityRef.ask(command(_))
                onSuccess(reply) {
                  case StatusReply.Success(summary: ShoppingCart.Summary) =>
                    complete(StatusCodes.OK -> summary)
                  case StatusReply.Error(reason) =>
                    complete(StatusCodes.BadRequest -> reason)
                }
            }
          },
          pathPrefix(Segment) { cartId =>
            concat(get {
              val entityRef =
                sharding.entityRefFor(ShoppingCart.EntityKey, cartId)
              onSuccess(entityRef.ask(ShoppingCart.Get)) { summary =>
                if (summary.items.isEmpty) complete(StatusCodes.NotFound)
                else complete(summary)
              }
            }, path("checkout") {
              post {
                val entityRef =
                  sharding.entityRefFor(ShoppingCart.EntityKey, cartId)
                val reply: Future[StatusReply[ShoppingCart.Summary]] =
                  entityRef.ask(ShoppingCart.Checkout(_))
                onSuccess(reply) {
                  case StatusReply.Success(summary: ShoppingCart.Summary) =>
                    complete(StatusCodes.OK -> summary)
                  case StatusReply.Error(reason) =>
                    complete(StatusCodes.BadRequest -> reason)
                }
              }
            })
          })
      }
    }

}

object JsonFormats {

  import spray.json.RootJsonFormat
  // import the default encoders for primitive types (Int, String, Lists etc)
  import spray.json.DefaultJsonProtocol._

  implicit val summaryFormat: RootJsonFormat[ShoppingCart.Summary] =
    jsonFormat2(ShoppingCart.Summary)
  implicit val addItemFormat: RootJsonFormat[ShoppingCartRoutes.AddItem] =
    jsonFormat3(ShoppingCartRoutes.AddItem)
  implicit val updateItemFormat: RootJsonFormat[ShoppingCartRoutes.UpdateItem] =
    jsonFormat3(ShoppingCartRoutes.UpdateItem)

}

File: akka-sample-cqrs-java/src/main/java/sample/cqrs/ShoppingCartRoutes.java
package sample.cqrs;

import akka.actor.typed.ActorSystem;
import akka.actor.typed.javadsl.Adapter;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.EntityRef;
import akka.http.javadsl.marshallers.jackson.Jackson;
import akka.http.javadsl.model.StatusCodes;
import akka.http.javadsl.server.PathMatchers;
import akka.http.javadsl.server.Route;
import akka.serialization.jackson.JacksonObjectMapperProvider;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.time.Duration;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

import static akka.http.javadsl.server.Directives.*;

public class ShoppingCartRoutes {

  public static class AddItem {
    public final String cartId;
    public final String itemId;
    public final int quantity;

    public AddItem(String cartId, String itemId, int quantity) {
      this.cartId = cartId;
      this.itemId = itemId;
      this.quantity = quantity;
    }
  }

  public static class UpdateItem {
    public final String cartId;
    public final String itemId;
    public final int quantity;

    public UpdateItem(String cartId, String itemId, int quantity) {
      this.cartId = cartId;
      this.itemId = itemId;
      this.quantity = quantity;
    }
  }

  private final ActorSystem<?> system;
  private final ClusterSharding sharding;
  private final Duration timeout;
  private final ObjectMapper objectMapper;

  public ShoppingCartRoutes(ActorSystem<?> system) {
    this.system = system;
    sharding = ClusterSharding.get(system);
    timeout = system.settings().config().getDuration("shopping.askTimeout");
    // Use Jackson ObjectMapper from Akka Jackson serialization
    objectMapper = JacksonObjectMapperProvider.get(Adapter.toClassic(system))
      .getOrCreate("jackson-json", Optional.empty());
  }


  public Route shopping() {
    return pathPrefix("shopping", () ->
      pathPrefix("carts", () ->
        concat(
          pathPrefix(PathMatchers.segment(), (String cartId) ->
            concat(
              get(() -> onSuccess(getCart(cartId), summary -> {
                if (summary.items.isEmpty())
                  return complete(StatusCodes.NOT_FOUND);
                else
                  return complete(StatusCodes.OK, summary, Jackson.marshaller(objectMapper));
              })),
              path("checkout", () ->
                post(() -> onConfirmationReply(checkout(cartId)))
              )
            )
          ),
          post(() ->
              entity(
                Jackson.unmarshaller(objectMapper, AddItem.class),
                data -> onConfirmationReply(addItem(data)))),
          put(() ->
            entity(
              Jackson.unmarshaller(objectMapper, UpdateItem.class),
              data -> onConfirmationReply(updateItem(data))))
        )
      )
    );
  }

  private Route onConfirmationReply(CompletionStage<ShoppingCart.Confirmation> reply) {
    return onSuccess(reply, confirmation -> {
      if (confirmation instanceof ShoppingCart.Accepted)
        return complete(StatusCodes.OK, ((ShoppingCart.Accepted) confirmation).summary, Jackson.marshaller(objectMapper));
      else
        return complete(StatusCodes.BAD_REQUEST, ((ShoppingCart.Rejected) confirmation).reason);
    });
  }

  private CompletionStage<ShoppingCart.Confirmation> addItem(AddItem data) {
    EntityRef<ShoppingCart.Command> entityRef =
      sharding.entityRefFor(ShoppingCart.ENTITY_TYPE_KEY, data.cartId);
    return entityRef.ask(replyTo -> new ShoppingCart.AddItem(data.itemId, data.quantity, replyTo), timeout);
  }

  private CompletionStage<ShoppingCart.Confirmation> updateItem(UpdateItem data) {
    EntityRef<ShoppingCart.Command> entityRef =
      sharding.entityRefFor(ShoppingCart.ENTITY_TYPE_KEY, data.cartId);
    if (data.quantity == 0)
      return entityRef.ask(replyTo -> new ShoppingCart.RemoveItem(data.itemId, replyTo), timeout);
    else
      return entityRef.ask(replyTo -> new ShoppingCart.AdjustItemQuantity(data.itemId, data.quantity, replyTo), timeout);
  }

  private CompletionStage<ShoppingCart.Summary> getCart(String cartId) {
    EntityRef<ShoppingCart.Command> entityRef =
      sharding.entityRefFor(ShoppingCart.ENTITY_TYPE_KEY, cartId);
    return entityRef.ask(ShoppingCart.Get::new, timeout);
  }

  private CompletionStage<ShoppingCart.Confirmation> checkout(String cartId) {
    EntityRef<ShoppingCart.Command> entityRef =
      sharding.entityRefFor(ShoppingCart.ENTITY_TYPE_KEY, cartId);
    return entityRef.ask(ShoppingCart.Checkout::new, timeout);
  }


}
